Implement PyNaCl backend for Ed25519 keys (part of RFC 8037)

blag opened the pull request:
[RFC8037](https://tools.ietf.org/html/rfc8037) is an extension to JOSE that includes Ed25519 and Ed448 keys.

This PR adds support for Ed25519 keys in a new `nacl` backend using the wonderful `PyNaCl` package, and integrates it into the JWK portion of this project.

Unfortunately, while Ed448 keys are part of RFC8037, there are no good Python libraries for those keys yet, so support for them was left out. Implementation of that is a good candidate for future work, but will not be considered part of this PR.

This PR should be good to go (assuming it passes CI tests). 😃

blag commented:
Oh, and I forgot to mention: all of the code that I added has 100% test coverage. 👍

pohutukawa commented:
I'd really like to see this land in `python-jose`, as this is straight on route towards also implementing ChaCha20/Poly1305 for JWE. I'd be willing to contribute, as I need both (JWT with Ed25519 and JWE with the above mentioned cipher).

mpdavis commented:
First, my apologies for taking so long to take a look at this. 

The only issue I see with this PR is that it introduces a C-based python dependency, which breaks the compatibility with Google App Engine. That said, GAE support has proven to be enough of a pain that I am likely to drop support for it entirely to avoid these complications trying to support further extensions of the JOSE framework.

I would like to officially drop support for GAE compatibility before merging this PR. I will see about getting that done.

mpdavis commented:
@blag I have successfully talked myself about worrying too much about getting a pure-python route for now. If we get these merge conflicts resolved, I can get this merged and simply cut a `4.0.0` release and let anyone that needs a pure-python deploy pin to 3.0.1 for now.

zejn commented:
I think that with a few small changes pure-Python can still be usable. And not having to install software you're not using is always a bonus.

First the Ed25519Key imports need to be made so they pass on ImportError both in jose/backends/__init__.py and jose/jwk.py.
Then advertising ED support in jose/constants.py must be made dependant on successful import of Ed25519Key.
Then it makes sense to add a new entry into extras_require so one can say `pip install python-jose[pynacl]` to enable Ed25519 support.

pohutukawa commented:
Are there any insights on whether this branch (after fixing the conflicts of course) will get merged?
It would mean that I can much more easily get a crack on stuff from #104, and thus discontinue my 'home brew' implementation of a *very* limited JOSE set.

blag commented on jose/constants.py:
```diff
@@ -12,12 +12,15 @@ class Algorithms(object):
+    ED = {EdDSA}
 
-    SUPPORTED = HMAC.union(RSA).union(EC)
+    SUPPORTED = HMAC.union(RSA).union(EC).union(ED)
```
I cannot import anything from `jose.backends` here, because all backends import from `jose.constants`, so `from jose.backends import Ed25519Key` here creates a circular import.

To work around this, the `nacl_backend` modifies the `ALGORITHMS.SUPPORTED` and `ALGORITHMS.ALL` variables directly. This is not ideal, but it's the only way (that I know of) to implement this in a decoupled manner, and the best way that supports Python < 3.6.

On that note, the [`__init_subclass__` hook](https://docs.python.org/3/reference/datamodel.html#object.__init_subclass__) is probably the best way to implement this in Python 3.6+.

blag commented:
I rebased this on `master` and implemented the changes @zejn recommended. Ed25519 support is now optional. I'm not super happy with how I had to workaround the circular imports issue with `jose.constants`, but it works.

This is ready for review.

pohutukawa commented:
@mpdavis, there's been a lot of work recently gone into sorting out crypto backend implementation details. Does this PR stand a chance of getting merged somewhere soonish?

blag commented:
@zejn and/or @mpdavis: Please review.

Mischala commented:
Any idea when this is likely to go in? 
Would be very useful for a project i am working on.
Thanks!

blag commented:
@zejn and @mpdavis - I have emailed both of you offering my assistance in maintaining this project. If you do not respond either privately via email or here by the end of next week, I will be forking this project and publishing my fork to PyPI.

Thank you both for your work so far!

mpdavis commented:
I think it is valid to be frustrated with my level of attention.

It makes sense to me to simply accept the offer of help instead of continuing trying to find free time to work on it, only to continue pushing it off. 

I have sent invites to both @zejn and @blag to collaborate on the project.

I do appreciate the offers of help, and I apologize for the frustrating lack of communication on my part.

pohutukawa commented:
It would be awesome if development could pick up (again) for this. Especially as it's called `python-jose`, but basically just does JWS/JWT. I'd like to see if I can contribute a starting set of JWE to this ensemble as well (based on my private/limited stand-alone implementation).

zejn commented:
This looks ok to me.

blag commented:
In light of https://github.com/pyca/pynacl/pull/507, I'm going to remove our dependence on encoders. I'll push a few more commits to this PR, get somebody to review the additional commits, then merge it.

blag commented:
This is again ready for review. None of my previous commits were changed, only 867f8fb and 9bcae9e were added.

blag commented:
@mpdavis Thank you for adding me as a collaborator, I very much appreciate it. I've had to make the call to stop maintaining other Python packages, so I understand how weighty of a decision it can be.

@zejn Sorry to string you along, but do you mind reviewing this one last time? I removed the reliance on PyNaCl's decoders since they are being removed in pyca/pynacl#507.

zejn commented on jose/constants.py:
```diff
@@ -37,3 +41,15 @@ class Algorithms(object):
+
+
+class Usages(object):
+    PUBLIC = 'public'
```
Looking at this again I was a bit puzzled why the need to specify usage. Turns out that both Ed25519 public and private keys are 32 bytes in length and there's no clear way to distinguish them. It seems that an extra parameter is a clumsy way to indicate the key type, with a potential for security issues, eg. by mixing up key usage.

Luckily, it also turns out that a PEM key format for Ed25519 keys was standardized in [RFC8410](https://tools.ietf.org/html/rfc8410) while this PR was open. Thus, I think a self describing format would make it harder to use keys incorrectly and would also make this implementation much more maintainable in the long run by not having to provide backwards compatibility by accepting raw keydata arguments.

Additionally, I think it would be beneficial to merge the mpdavis:backend-explicit-tests branch into master before merging this since there's been a lot of work done to make the backends independent and this is yet another backend. That branch also makes python-jose depend directly on pyasn1, which would be needed to implement PEM (de)serialization for this backend.

blag replied:
Good catch! I’m a little nervous that it’s still considered a draft specification, but I like the fact that our users would not need to explicitly specify the usage.

Your merge order seems sound to me, I’ll let you handle that merge since you’re already familiar with it and then I’ll deal with the rebase/merge fallout from that and implement reading from PEM files.

I’m not super happy about merging support for a specification that is in draft status, but I’ll try to keep an eye on that specification and keep this PR up to date with it. I’m also not sure how much proposed IETF standards change from draft to approval, and the only errata that has been reported so far is incredibly trivial. If that means we cut a release without this merged in, that’s alright.

mattsb42-aws replied:
@blag @zejn

I finally had some time to sit down with this, and I wanted to try and get some clarity on what exactly the ask is here.

Is there anything about JOSE that needs the RFC8410 encoding specifically, or are you just looking for *some* self-describing format? Looking at the way that the keys are loaded from the JOSE payload, it looks like they are only ever expected in "raw" format. Is this correct?

If you are just looking for *a* self-describing format and don't specifically need RFC8410, then especially since RFC8410 has yet to be finalized, I would like to propose an alternate approach:

As of v2.6, `pyca/cryptography` supports Ed25519 signing and verification[1]. It looks like this PR was originally cut before that was released (2019-02-27). While pyca/cryptography does not support RFC8410 encoding, and likely will not until OpenSSL does, it does support PKCS8 encoding of Ed25519 keys, which does give you the self-describing nature that you are looking for.

Replacing the `PyNaCl` use in this backend with `pyca/cryptography` use should be pretty straightforward, and that would let us avoid adding another leaf to the already complex dependency graph.

As for the explicit "use" flag in the API, a simpler approach might be to have the constructor expect an already loaded key and have separate `classmethod`s for loading a private key vs public key from "raw" encoded bytes.

[1] https://cryptography.io/en/latest/changelog/#v2-6

mattsb42-aws replied:
Any thoughts on this @blag @zejn @mpdavis ?

blag replied:
Thanks for digging into this @mattsb42-aws, I really appreciate it.

I think I'll close this PR and release a new version of python-jose later this week from `master`. That way we at least get `@zejn`'s patches out the door.

Once that's done I'll dig into getting Ed25519 support with the Cryptography library.

How does that sound?

pohutukawa replied:
Hmmm, Cryptography pretty much only offers Ed25519 (and some others for EdDSA), but no X25519 or other crypto primitives that could be used for actually enabling a "JOSE" library to do any more than JWS/JWT (e.g. using X25519 for ECDH for JWE).

So if someone was to asked me, my **very strong** preference for addressing this problem is still to provide a PyNaCl back-end, which would (down the line) resolve many more issues and enable further extensions not quite possible with Cryptography only (e.g. also modern and well accepted ciphers such as ChaCha20/Poly1305).

Just my NZ$0.02 on that topic ...

pohutukawa replied:
Addendum: I've just seen that Cryptography may have pulled even with providing ChaCha20/Poly1305 and X25519. So from that angle it would be OK to use Cryptography instead of PyNaCl.

My only preference for PyNaCl resides in the fact that its implementations as well as bindings to `libsodium` have been long tested and found to be good and sufficiently tested and performant.

mattsb42-aws replied:
I don't really have a stake in this, but from my perspective I'd be happy with either direction. PyNaCl and pyca/cryptography are owned and maintained by mostly the same people, and libsodium and libcrypto are both well established cryptographic codebases.

The main reason I would argue for using pyca/cryptography is to reduce the number of backends.

pohutukawa replied:
The problem is not the source/author of the package (`pyca/cryptography` vs. `pyca/pynacl`), it's the source of the implementation of the ciphers. Cryptography used OpenSSL, which is known to be quite problematic. Its code base is not very good, thus the reasons for all the forks (e.g. LibreSSL, BoringSSL and others), particularly after the Heartbleed "incident".

In contrast libsodium was **always** very tight, secure and tidy.

From a functionality perspective though, I'd say you can go with Cryptography ...

blag commented:
As per [this conversation](https://github.com/mpdavis/python-jose/pull/100#pullrequestreview-223591449), I'm closing this PR.
